pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py

```diff
@@ -133,6 +133,85 @@
 _BIDIRECTIONAL_CONSUMER_NAME = 'Thread-ConsumeBidirectionalStream'
 
 
+class _RequestQueueGenerator(object):
+    """A helper for sending requests to a gRPC stream from a Queue.
+
+    This generator takes requests off a given queue and yields them to gRPC.
+
+    This helper is useful when you have an indeterminate, indefinite, or
+    otherwise open-ended set of requests to send through a request-streaming
+    (or bidirectional) RPC.
+
+    The reason this is necessary is because gRPC takes an iterator as the
+    request for request-streaming RPCs. gRPC consumes this iterator in another
+    thread to allow it to block while generating requests for the stream.
+    However, if the generator blocks indefinitely gRPC will not be able to
+    clean up the thread as it'll be blocked on `next(iterator)` and not be able
+    to check the channel status to stop iterating. This helper mitigates that
+    by waiting on the queue with a timeout and checking the RPC state before
+    yielding.
+
+    Finally, it allows for retrying without swapping queues because if it does
+    pull an item off the queue, it'll immediately put it back and then exit.
+    This is necessary because yielding the item in this case will cause gRPC
+    to discard it. In practice, this means that the order of messages is not
+    guaranteed. If such a thing is necessary it would be easy to use a priority
+    queue.
+
+    Example::
+
+        requests = request_queue_generator(q)
+        rpc = stub.StreamingRequest(iter(requests))
+        requests.rpc = rpc
+
+        for response in rpc:
+            print(response)
+            q.put(...)
+    """
+    def __init__(self, queue, period=1, initial_request=None):
+        self._queue = queue
+        self._period = period
+        self._initial_request = initial_request
+        self.rpc = None
+
+    def _should_exit(self):
+        # Note: there is a possibility that this starts *before* the rpc
+        # property is set. So we have to check if self.rpc is set before seeing
+        # if it's active.
+        if self.rpc is not None and not self.rpc.is_active():
+            return True
+        else:
+            return False
```

Comment:
In the "racy" case indicated by the comment, how do we expect the user to recover from this state?

Reply:
It's okay that it's racy. The resolution is to just keep the generator open. The idea is that there are three states:

1. Generator created, rpc not set.
2. gRPC started a thread to consume the generator, rpc is still not set.
3. rpc is set, now start monitoring it to see when it stops.

So in all of those cases the right thing to do is to just keep the generator open.